central/group/datastore/internal/store/store_impl.go

```diff
@@ -146,28 +135,119 @@ func (s *storeImpl) Mutate(toRemove, toUpdate, toAdd []*storage.Group) error {
+
+	// TODO(ROX-11592): Once the deprecation of retrieving groups by their properties is fully deprecated, this condition
+	// can be removed and groups shall only be retrievable via their id.
+	if id != "" {
+		if buc.Get([]byte(id)) == nil {
+			return errox.NotFound.Newf("group config for %q does not exist", id)
+		}
+	} else {
+		grp, err := getByPropsInTransaction(tx, group.GetProps())
+		if err != nil {
+			return err
+		}
+		if grp == nil {
+			return errox.NotFound.Newf("group config for (auth provider id=%s, key=%s, value=%s) does not exist",
+				group.GetProps().GetAuthProviderId(), group.GetProps().GetKey(), group.GetProps().GetValue())
+		}
+		id = grp.GetProps().GetId()
 	}
-	return buc.Put(key, value)
+
+	bytes, err := proto.Marshal(group)
+	if err != nil {
+		return errox.InvariantViolation.CausedBy(err)
+	}
+
+	return buc.Put([]byte(id), bytes)
 }
 
-func removeInTransaction(tx *bolt.Tx, key []byte) error {
+func removeInTransaction(tx *bolt.Tx, props *storage.GroupProperties) error {
 	buc := tx.Bucket(groupsBucket)
-	if buc.Get(key) == nil {
-		return errox.NotFound.Newf("group config for %q does not exist", key)
+	id := props.GetId()
+
+	// TODO(ROX-11592): Once the deprecation of retrieving groups by their properties is fully deprecated, this condition
+	// can be removed and groups shall only be retrievable via their id.
+	if id != "" {
+		if buc.Get([]byte(id)) == nil {
+			return errox.NotFound.Newf("group config for %q does not exist", id)
+		}
+	} else {
+		grp, err := getByPropsInTransaction(tx, props)
+		if err != nil {
+			return err
+		}
+		if grp == nil {
+			return errox.NotFound.Newf("group config for (auth provider id=%s, key=%s, value=%s) does not exist",
+				props.GetAuthProviderId(), props.GetKey(), props.GetValue())
+		}
+		id = grp.GetProps().GetId()
 	}
-	return buc.Delete(key)
+
+	return buc.Delete([]byte(id))
+}
+
+func filterInTransaction(tx *bolt.Tx, filter func(*storage.GroupProperties) bool) (grps []*storage.Group, err error) {
+	buc := tx.Bucket(groupsBucket)
+
+	err = buc.ForEach(func(k, v []byte) error {
+		var grp storage.Group
+		err := proto.Unmarshal(v, &grp)
+		if err != nil {
+			return err
+		}
+		if filter == nil || filter(grp.GetProps()) {
+			grps = append(grps, &grp)
+		}
+		return nil
+	})
+	return grps, err
+}
+
+func filterByPropsInTransaction(tx *bolt.Tx, props *storage.GroupProperties) ([]*storage.Group, error) {
+	grps, err := filterInTransaction(tx, func(stored *storage.GroupProperties) bool {
+		if props.GetAuthProviderId() != stored.GetAuthProviderId() ||
+			props.GetKey() != stored.GetKey() ||
+			props.GetValue() != stored.GetValue() {
+			return false
+		}
+		return true
+	})
+	if err != nil {
+		return nil, err
+	}
+	return grps, nil
+}
+
+func getByPropsInTransaction(tx *bolt.Tx, props *storage.GroupProperties) (*storage.Group, error) {
+	grps, err := filterByPropsInTransaction(tx, props)
+	if err != nil {
+		return nil, err
+	}
+	if len(grps) == 0 {
```

Comment:
You previously had an error case if more than one group has been found. I think you want to put it back here.
```
	if len(grps) > 1 {
		return nil, errox.InvalidArgs.Newf("multiple groups found for properties (auth provider id=%s, key=%s, "+
			"value=%s), provide an ID to retrieve a group unambiguously",
			props.GetAuthProviderId(), props.GetKey(), props.GetValue())
```